--- spice-gtk-0.31/src/usbutil.c.orig	2015-11-11 22:28:41.000000000 +0100
+++ spice-gtk-0.31/src/usbutil.c	2020-05-16 22:40:00.116238729 +0200
@@ -25,6 +25,12 @@
 #include <glib/gi18n.h>
 #include <ctype.h>
 #include <stdlib.h>
+#ifdef __linux__
+#include <sys/sysmacros.h>
+#ifndef major /* major and minor macros were moved to sys/sysmacros.h from sys/types.h */
+#include <sys/types.h>
+#endif
+#endif
 
 #include "glib-compat.h"
 
